org-code-javabuilder/validation/src/main/java/org/code/validation/NeighborhoodLog.java

```diff
@@ -0,0 +1,96 @@
+package org.code.validation;
+
+public class NeighborhoodLog {
+  private PainterLog[] painterLogs;
+  private String[][] finalOutput;
+
+  public NeighborhoodLog(PainterLog[] painterLogs, String[][] finalOutput) {
+    this.painterLogs = painterLogs;
+    this.finalOutput = finalOutput;
+  }
+
+  public PainterLog[] getPainterLogs() {
+    return this.painterLogs;
+  }
+
+  public String[][] getFinalOutput() {
+    return this.finalOutput;
+  }
+
+  /**
+   * @param eventType
+   * @param times
+   * @return true if the given eventType occurred exactly "times" times for at least one painter,
+   *     false otherwise.
+   */
+  public boolean onePainterDidAction(EventType eventType, int times) {
+    for (PainterLog painterLog : this.painterLogs) {
+      if (painterLog.didActionExactly(eventType, times)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /**
+   * @param eventType
+   * @param times
+   * @return true if the given eventType occurred exactly "times" times across all painters, false
+   *     otherwise.
+   */
+  public boolean actionHappened(EventType eventType, int times) {
+    int actionCount = 0;
+    for (PainterLog painterLog : this.painterLogs) {
+      actionCount += painterLog.actionCount(eventType);
```

Comment:
Do we need to worry about infinite loops or large inputs? i.e. if times = 1000000 or the student code loops thousands of times? (Not sure that we do - curious for your thoughts)

Reply:
I think in those cases a couple things could happen: the code could be slow, but finish, we could time out, or we could hit some out of memory error. If we want student code to still pass then we would want to find a workaround, but my gut feeling is that would be a case where their code wouldn't pass, as they likely wouldn't even be able to run their code without timing out.